pyfarm/agent/service.py

```diff
@@ -222,83 +217,95 @@ def repeating_call(
 
     def should_reannounce(self):
         """Small method which acts as a trigger for :meth:`reannounce`"""
-        if self.reannounce_client_request is not None:
-            svclog.debug("Agent is already trying to announce itself.")
-            return
-
-        if self.shutting_down:
+        if self.reannouce_lock.locked or self.shutting_down:
             return False
 
         contacted = config.master_contacted(update=False)
-        remaining = (datetime.utcnow() - contacted).total_seconds()
-        return remaining > config["agent_master_reannounce"]
+        if contacted is None:
+            return True
+
+        return utility.total_seconds(
+            datetime.utcnow() - contacted) > config["agent_master_reannounce"]
 
+    @inlineCallbacks
     def reannounce(self, force=False):
         """
         Method which is used to periodically contact the master.  This
         method is generally called as part of a scheduled task.
         """
+        yield self.reannouce_lock.acquire()
```

Comment:
Good point, even with the should_reannounce() check it could still be an issue.  I'l look into adding this, probably just need to override acquire() and raise a timeout exception of some kind.